File: npm/packages/kge/bench/lib/adversarial.mjs
// Adversarial symmetry-decoy generator (ADR-005 threat 1; ADR-006 §protocol 5).
// A JS re-implementation, for the harness's OWN use, of the symmetry-pattern
// poisoning attack (Bhardwaj et al., ACL-IJCNLP 2021, arXiv:2111.06345): decoy
// triples added elsewhere in the graph indirectly shift confidence on a target
// fact. Symmetry patterns generalize across every model/dataset the paper
// tested, which is why they are the CI regression vector.
//
// Simplification vs the paper: the paper selects decoy entities by a
// soft-truth / influence ranking; here decoy tails are chosen deterministically
// among entities not already linked to the anchor (selection simplified,
// pattern preserved). The gate is that mean confidence on the targeted facts
// DROPS under attack (ADR-006), not that the fact flips.

import { mulberry32 } from './metrics.mjs';

const triKey = (s, r, o) => `${s} ${r} ${o}`;

/**
 * Detect (near-)symmetric relations by reciprocal fraction: the share of
 * (s,r,o) triples (s≠o) whose reverse (o,r,s) is also present. A fully
 * reciprocal relation scores 1.0; a directed one scores ~0.
 *
 * @param triples  array of {s,r,o}
 * @param opts.threshold reciprocal fraction to call a relation symmetric (0.8)
 * @param opts.minCount  ignore relations with fewer than this many triples (10)
 * @returns array of { relation, fraction, count } sorted by fraction desc
 */
export function detectSymmetricRelations(triples, { threshold = 0.8, minCount = 10 } = {}) {
  const present = new Set(triples.map((t) => triKey(t.s, t.r, t.o)));
  const total = new Map();
  const recip = new Map();
  for (const t of triples) {
    if (t.s === t.o) continue;
    total.set(t.r, (total.get(t.r) ?? 0) + 1);
    if (present.has(triKey(t.o, t.r, t.s))) recip.set(t.r, (recip.get(t.r) ?? 0) + 1);
  }
  const out = [];
  for (const [r, n] of total) {
    if (n < minCount) continue;
    const fraction = (recip.get(r) ?? 0) / n;
    out.push({ relation: r, fraction, count: n, symmetric: fraction >= threshold });
  }
  return out.sort((a, b) => b.fraction - a.fraction);
}

/**
 * Detect inverse relation pairs by the share of (s,r1,o) whose (o,r2,s) exists.
 * Reported informationally in the receipt; not gated.
 * @returns array of { relation, inverse, fraction, count }
 */
export function detectInversePairs(triples, { threshold = 0.8, minCount = 10 } = {}) {
  const present = new Set(triples.map((t) => triKey(t.s, t.r, t.o)));
  const byRel = new Map();
  for (const t of triples) {
    if (t.s === t.o) continue;
    if (!byRel.has(t.r)) byRel.set(t.r, []);
    byRel.get(t.r).push(t);
  }
  const rels = [...byRel.keys()];
  const out = [];
  for (const r1 of rels) {
    const items = byRel.get(r1);
    const hits = new Map();
    for (const t of items) {
      for (const r2 of rels) {
        if (r2 === r1) continue;
        if (present.has(triKey(t.o, r2, t.s))) hits.set(r2, (hits.get(r2) ?? 0) + 1);
      }
    }
    for (const [r2, h] of hits) {
      const fraction = h / items.length;
      if (fraction >= threshold && items.length >= minCount) {
        out.push({ relation: r1, inverse: r2, fraction, count: items.length });
      }
    }
  }
  return out;
}

/**
 * Pick target facts to attack: test triples under a symmetric relation (the
 * generalizing vector). Deterministic: sorted by triple key, first `count`.
 */
export function pickTargets(testTriples, symmetricRelations, count = 20) {
  const symSet = new Set(symmetricRelations.filter((r) => r.symmetric).map((r) => r.relation));
  const eligible = testTriples
    .filter((t) => symSet.has(t.r) && t.s !== t.o)
    .sort((a, b) => triKey(a.s, a.r, a.o).localeCompare(triKey(b.s, b.r, b.o)));
  return eligible.slice(0, count);
}

/**
 * Generate symmetry decoys for the targets. For target (s,r,o) emit
 * `perTarget` decoys (o, r, s′) with s′ chosen among entities not already an
 * r-tail of o (and ≠ o, s), deterministically. Never emits a triple already in
 * `existing`.
 *
 * @param targets    array of {s,r,o}
 * @param entities   full entity id list
 * @param existing   array of {s,r,o} already in the graph
 * @param opts.perTarget decoys per target (default 3)
 * @param opts.seed  PRNG seed (default 1337)
 * @returns array of {s,r,o} decoy triples (deduped, disjoint from existing)
 */
export function generateSymmetryDecoys(targets, entities, existing, { perTarget = 3, seed = 1337 } = {}) {
  const present = new Set(existing.map((t) => triKey(t.s, t.r, t.o)));
  const tailsOf = new Map(); // `${o}|${r}` -> Set of existing tails
  for (const t of existing) {
    const key = `${t.o}|${t.r}`;
    if (!tailsOf.has(key)) tailsOf.set(key, new Set());
    tailsOf.get(key).add(t.s); // subjects that already point r->o
  }
  const rng = mulberry32(seed);
  const decoys = [];
  for (const t of targets) {
    const linked = tailsOf.get(`${t.o}|${t.r}`) ?? new Set();
    let made = 0;
    let attempts = 0;
    while (made < perTarget && attempts < entities.length * 2) {
      attempts++;
      const sPrime = entities[Math.floor(rng() * entities.length)];
      if (sPrime === t.o || sPrime === t.s) continue;
      if (linked.has(sPrime)) continue;
      const k = triKey(t.o, t.r, sPrime);
      if (present.has(k)) continue;
      present.add(k);
      linked.add(sPrime);
      decoys.push({ s: t.o, r: t.r, o: sPrime });
      made++;
    }
  }
  return decoys;
}

File: crates/ruvector-kge/src/adversarial.rs
//! Adversarial and security checks (ADR-005). Three things live here: the
//! symmetry-pattern decoy-triple generator (the primary poisoning CI case), a
//! membership-inference probe, and the input-limit guard the bindings call
//! before the FFI boundary. None of these ever logs triple text (ADR-005).

use crate::{EntityId, KgeError, RelationId, Result, Triple};
use std::collections::HashSet;

// ---- Input limits (ADR-005) ------------------------------------------------

/// ADR-005 table limits (index-build budget).
pub const MAX_ENTITIES: usize = 1_000_000;
/// ADR-005 table limits.
pub const MAX_RELATIONS: usize = 100_000;
/// Local defensive bound — ADR-005 states no explicit triple cap; this guards
/// the ingest path from an unbounded edge list.
pub const MAX_TRIPLES: usize = 100_000_000;
/// Local defensive bound — ADR-005 states no explicit dimension cap; this keeps
/// the ANN index build within budget.
pub const MAX_DIMS: usize = 4096;

/// Reject an oversized table before it reaches the scorer/index (ADR-005:
/// rejected with a typed error, never silently truncated).
pub fn check_limits(
    num_entities: usize,
    num_relations: usize,
    num_triples: usize,
    dims: usize,
) -> Result<()> {
    if num_entities > MAX_ENTITIES {
        return Err(KgeError::Limit("more than 1M entities"));
    }
    if num_relations > MAX_RELATIONS {
        return Err(KgeError::Limit("more than 100k relations"));
    }
    if num_triples > MAX_TRIPLES {
        return Err(KgeError::Limit("more than 100M triples"));
    }
    if dims == 0 || dims > MAX_DIMS {
        return Err(KgeError::Limit("dims out of range (1..=4096)"));
    }
    Ok(())
}

// ---- Symmetry-pattern decoy poisoning (Bhardwaj 2021) ----------------------

fn xorshift64(state: &mut u64) -> u64 {
    let mut x = *state;
    x ^= x << 13;
    x ^= x >> 7;
    x ^= x << 17;
    *state = x;
    x
}

/// Generate up to `n` symmetry-pattern decoy triples targeting `relation`
/// (Bhardwaj et al., ACL-IJCNLP 2021, arXiv:2111.06345 — the attack that
/// generalises across every model/dataset tested).
///
/// The pattern: for a relation the model has learned as **symmetric**, a target
/// fact `(s, r, o)` is attacked *indirectly*. Adding `(o, r, s')` for a
/// distractor `s' ∉ {s, o}` makes symmetry infer `(s', r, o)` — a competitor
/// that steals rank and score mass from the target subject `s` without the
/// attacker ever touching `(s, r, o)` itself. We deliberately do **not** emit
/// the target's own inverse `(o, r, s)`: for a symmetric relation that
/// *reinforces* the target rather than poisoning it.
///
/// Decoys are deduplicated against `store` and against each other, and capped
/// at `n`. `seed` makes the distractor choice deterministic. The entity id
/// space is inferred from `store` (max id + 1).
#[must_use]
pub fn symmetry_decoys(store: &[Triple], relation: RelationId, n: usize, seed: u64) -> Vec<Triple> {
    let targets: Vec<&Triple> = store.iter().filter(|t| t.r == relation).collect();
    if targets.is_empty() || n == 0 {
        return Vec::new();
    }
    let num_entities = store
        .iter()
        .flat_map(|t| [t.s, t.o])
        .max()
        .map_or(0, |m| m as u64 + 1);
    if num_entities < 3 {
        return Vec::new();
    }
    let existing: HashSet<Triple> = store.iter().copied().collect();
    let mut seen: HashSet<Triple> = HashSet::new();
    let mut out = Vec::with_capacity(n);
    let mut state = seed | 1;

    let mut i = 0usize;
    // Bounded scan: at most n·64 attempts so a saturated graph terminates.
    let max_attempts = n.saturating_mul(64).max(n);
    let mut attempts = 0usize;
    while out.len() < n && attempts < max_attempts {
        attempts += 1;
        let target = targets[i % targets.len()];
        i += 1;
        // Pick a distractor s' ∉ {s, o}.
        let mut sprime = (xorshift64(&mut state) % num_entities) as EntityId;
        if sprime == target.s || sprime == target.o {
            sprime = ((sprime as u64 + 1) % num_entities) as EntityId;
        }
        if sprime == target.s || sprime == target.o {
            continue; // tiny graph collision; try another target/seed draw
        }
        let decoy = Triple::new(target.o, relation, sprime);
        if existing.contains(&decoy) || !seen.insert(decoy) {
            continue;
        }
        out.push(decoy);
    }
    out
}

/// Per-target confidence drop under an attack. The gate (ADR-006) is that
/// confidence **drops** under attack, not that accuracy is perfect.
#[derive(Debug, Clone, PartialEq)]
pub struct AttackReport {
    pub n: usize,
    pub mean_before: f32,
    pub mean_after: f32,
    /// `mean(before − after)` — positive means the attack degraded confidence.
    pub mean_drop: f32,
    pub max_drop: f32,
    /// Share of targeted facts whose score decreased.
    pub fraction_dropped: f32,
    /// `true` when mean confidence dropped (the honest-degradation signal).
    pub dropped: bool,
}

/// Compare targeted-triple scores before and after an attack. `before` and
/// `after` are paired per target; the shorter length wins if they differ.
#[must_use]
pub fn attack_report(before: &[f32], after: &[f32]) -> AttackReport {
    let n = before.len().min(after.len());
    if n == 0 {
        return AttackReport {
            n: 0,
            mean_before: 0.0,
            mean_after: 0.0,
            mean_drop: 0.0,
            max_drop: 0.0,
            fraction_dropped: 0.0,
            dropped: false,
        };
    }
    let mut sum_before = 0.0f32;
    let mut sum_after = 0.0f32;
    let mut max_drop = f32::NEG_INFINITY;
    let mut dropped_count = 0usize;
    for i in 0..n {
        let (b, a) = (before[i], after[i]);
        sum_before += b;
        sum_after += a;
        let drop = b - a;
        if drop > max_drop {
            max_drop = drop;
        }
        if a < b {
            dropped_count += 1;
        }
    }
    let mean_before = sum_before / n as f32;
    let mean_after = sum_after / n as f32;
    let mean_drop = mean_before - mean_after;
    AttackReport {
        n,
        mean_before,
        mean_after,
        mean_drop,
        max_drop,
        fraction_dropped: dropped_count as f32 / n as f32,
        dropped: mean_drop > 0.0,
    }
}

// ---- Membership inference (Wang & Sun, arXiv:2104.08273) --------------------

/// Score-gap statistic between train and held-out triples. A large positive gap
/// means a query-only attacker can infer training-set membership.
#[derive(Debug, Clone, PartialEq)]
pub struct MembershipReport {
    pub train_mean: f32,
    pub holdout_mean: f32,
    /// `train_mean − holdout_mean`.
    pub gap: f32,
    pub threshold: f32,
    /// `true` when the gap exceeds `threshold` (membership is inferable).
    pub alarm: bool,
}

/// Probe membership-inference leakage: the gap between mean train and mean
/// held-out scores, alarming past `threshold` (ADR-005 medium/high threat).
#[must_use]
pub fn membership_inference_probe(
    train_scores: &[f32],
    holdout_scores: &[f32],
    threshold: f32,
) -> MembershipReport {
    let train_mean = mean(train_scores);
    let holdout_mean = mean(holdout_scores);
    let gap = train_mean - holdout_mean;
    MembershipReport {
        train_mean,
        holdout_mean,
        gap,
        threshold,
        alarm: gap > threshold,
    }
}

fn mean(xs: &[f32]) -> f32 {
    if xs.is_empty() {
        0.0
    } else {
        xs.iter().sum::<f32>() / xs.len() as f32
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn check_limits_rejects_oversized_tables() {
        assert!(check_limits(10, 10, 100, 128).is_ok());
        assert!(matches!(
            check_limits(MAX_ENTITIES + 1, 10, 100, 128),
            Err(KgeError::Limit(_))
        ));
        assert!(matches!(
            check_limits(10, MAX_RELATIONS + 1, 100, 128),
            Err(KgeError::Limit(_))
        ));
        assert!(matches!(
            check_limits(10, 10, 100, 0),
            Err(KgeError::Limit(_))
        ));
        assert!(matches!(
            check_limits(10, 10, 100, MAX_DIMS + 1),
            Err(KgeError::Limit(_))
        ));
    }

    #[test]
    fn decoys_exploit_symmetry_without_touching_the_target() {
        // A synthetic graph where relation 0 is symmetric: (s,0,o) & (o,0,s).
        let store = vec![
            Triple::new(0, 0, 1),
            Triple::new(1, 0, 0),
            Triple::new(2, 0, 3),
            Triple::new(3, 0, 2),
            Triple::new(4, 0, 5),
        ];
        let decoys = symmetry_decoys(&store, 0, 4, 12345);
        assert!(!decoys.is_empty());
        assert!(decoys.len() <= 4);
        for d in &decoys {
            // Every decoy uses the attacked relation.
            assert_eq!(d.r, 0);
            // A decoy is (o, r, s') — never a verbatim copy of an existing triple.
            assert!(!store.contains(d));
            // The distractor s' is not the object it hangs off.
            assert_ne!(d.s, d.o);
            // Structural symmetry property: the decoy's subject is the object of
            // some real target triple, so symmetry infers a competitor for it.
            assert!(store.iter().any(|t| t.r == 0 && t.o == d.s));
        }
        // Deterministic under a fixed seed.
        assert_eq!(decoys, symmetry_decoys(&store, 0, 4, 12345));
    }

    #[test]
    fn decoys_empty_for_absent_relation() {
        let store = vec![Triple::new(0, 0, 1), Triple::new(1, 0, 2)];
        assert!(symmetry_decoys(&store, 9, 4, 1).is_empty());
    }

    #[test]
    fn attack_report_shows_confidence_drop() {
        let before = [0.9, 0.8, 0.95, 0.7];
        let after = [0.5, 0.6, 0.4, 0.65];
        let r = attack_report(&before, &after);
        assert!(r.dropped);
        assert!(r.mean_drop > 0.0);
        assert_eq!(r.fraction_dropped, 1.0);
        assert!(r.max_drop >= 0.55 - 1e-6);
    }

    #[test]
    fn attack_report_no_drop_when_scores_hold() {
        let before = [0.5, 0.5];
        let after = [0.5, 0.6];
        let r = attack_report(&before, &after);
        assert!(!r.dropped);
    }

    #[test]
    fn membership_probe_alarms_on_large_gap() {
        let train = [0.95, 0.92, 0.98];
        let holdout = [0.40, 0.45, 0.38];
        let r = membership_inference_probe(&train, &holdout, 0.2);
        assert!(r.alarm);
        assert!(r.gap > 0.2);

        let tight = membership_inference_probe(&train, &train, 0.2);
        assert!(!tight.alarm);
    }
}
